pkg/bgpv1/metrics/metrics.go

```diff
@@ -0,0 +1,151 @@
+	"context"
+	"net/netip"
+	"strconv"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/cilium/cilium/pkg/bgpv1/agent"
+	"github.com/cilium/cilium/pkg/bgpv1/types"
+	"github.com/cilium/cilium/pkg/hive/cell"
+	"github.com/cilium/cilium/pkg/metrics"
+	"github.com/cilium/cilium/pkg/option"
+	"github.com/cilium/cilium/pkg/time"
+)
+
+const (
+	LabelVRouter  = "vrouter"
+	LabelNeighbor = "neighbor"
+	LabelAfi      = "afi"
+	LabelSafi     = "safi"
+
+	metricsSubsystem = "bgp_control_plane"
+)
+
+type collector struct {
+	SessionState          *prometheus.Desc
+	TotalAdvertisedRoutes *prometheus.Desc
+	TotalReceivedRoutes   *prometheus.Desc
+
+	in collectorIn
+}
+
+type collectorIn struct {
+	cell.In
+
+	DaemonConfig  *option.DaemonConfig
+	Registry      *metrics.Registry
+	RouterManager agent.BGPRouterManager
+}
+
+// RegisterCollector registers the BGP Control Plane metrics collector to the
+// global prometheus registry. We don't rely on the cell.Metric because the
+// collectors we can provide through cell.Metric needs to implement
+// prometheus.Collector per metric which is not optimal in our case. We can
+// retrieve the multiple metrics from the single call to
+// RouterManager.GetPeers() and it is wastefil to call the same function
+// multiple times for each metric. Thus, we provide a raw Collector through
+// MustRegister interface. We may want to revisit this in the future.
+func RegisterCollector(in collectorIn) {
+	// Don't provide the collector if BGP control plane is disabled
+	if !in.DaemonConfig.EnableBGPControlPlane {
+		return
+	}
+	in.Registry.MustRegister(&collector{
+		SessionState: prometheus.NewDesc(
+			prometheus.BuildFQName(metrics.Namespace, metricsSubsystem, "session_state"),
+			"Current state of the BGP session with the peer, Up = 1 or Down = 0",
+			[]string{LabelVRouter, LabelNeighbor}, nil,
+		),
+		TotalAdvertisedRoutes: prometheus.NewDesc(
+			prometheus.BuildFQName(metrics.Namespace, metricsSubsystem, "advertised_routes"),
+			"Total number of routes advertised to the peer",
+			[]string{LabelVRouter, LabelNeighbor, LabelAfi, LabelSafi}, nil,
+		),
+		TotalReceivedRoutes: prometheus.NewDesc(
+			prometheus.BuildFQName(metrics.Namespace, metricsSubsystem, "received_routes"),
+			"Total number of routes received from the peer",
+			[]string{LabelVRouter, LabelNeighbor, LabelAfi, LabelSafi}, nil,
+		),
+		in: in,
+	})
+}
+
+func (c *collector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- c.SessionState
+	ch <- c.TotalAdvertisedRoutes
+	ch <- c.TotalReceivedRoutes
+}
+
+func (c *collector) Collect(ch chan<- prometheus.Metric) {
+	// We defensively set a 1 min timeout here. When the underlying
+	// router is not responsive, we cannot make a progress. 1 min
+	// is chosen to be a too long time that we should never hit for
+	// normal cases.
+	ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
+	peers, err := c.in.RouterManager.GetPeers(ctx)
+	cancel()
+	if err != nil {
+		return
```

Comment:
Should we log an error / warning here to provide user some indication that the metrics may be out of date?